dtfs-central-api/src/services/state-machines/utilisation-report/event-handlers/add-a-payment/add-a-payment.event-handler.ts

```diff
@@ -0,0 +1,50 @@
+import { EntityManager } from 'typeorm';
+import { DbRequestSource, FeeRecordEntity, FeeRecordStatus, PaymentEntity, UtilisationReportEntity } from '@ukef/dtfs2-common';
+import { NewPaymentDetails } from '../../../../../types/utilisation-reports';
+import { BaseUtilisationReportEvent } from '../../event/base-utilisation-report.event';
+import { FeeRecordStateMachine } from '../../../fee-record/fee-record.state-machine';
+import { feeRecordsMatchAttachedPayments } from './helpers';
+
+type AddAPaymentEventPayload = {
+  transactionEntityManager: EntityManager;
+  feeRecords: FeeRecordEntity[];
+  paymentDetails: NewPaymentDetails;
+  requestSource: DbRequestSource;
+};
+
+export type UtilisationReportAddAPaymentEvent = BaseUtilisationReportEvent<'ADD_A_PAYMENT', AddAPaymentEventPayload>;
+
+export const handleUtilisationReportAddAPaymentEvent = async (
+  report: UtilisationReportEntity,
+  { transactionEntityManager, feeRecords, paymentDetails, requestSource }: AddAPaymentEventPayload,
+): Promise<UtilisationReportEntity> => {
+  const payment = PaymentEntity.create({
+    ...paymentDetails,
+    feeRecords,
+    requestSource,
+  });
+  await transactionEntityManager.save(PaymentEntity, payment);
+
+  const feeRecordStatusToSet: FeeRecordStatus = (await feeRecordsMatchAttachedPayments(feeRecords, transactionEntityManager)) ? 'MATCH' : 'DOES_NOT_MATCH';
+
+  const feeRecordStateMachines = feeRecords.map((feeRecord) => FeeRecordStateMachine.forFeeRecord(feeRecord));
+  await Promise.all(
+    feeRecordStateMachines.map((stateMachine) =>
+      stateMachine.handleEvent({
+        type: 'PAYMENT_ADDED',
+        payload: {
+          transactionEntityManager,
+          status: feeRecordStatusToSet,
+          requestSource,
+        },
+      }),
+    ),
+  );
```

Comment:
It feels out of place for the utilisation report state machine to decide the fee record status to set, I'd almost be tempted to not bother having a fee record state machine and let it set the statuses itself if this logic is already here. Instead, could our fee record state machine perhaps accept a list of fee records and then the logic for determining the status to set can live there? This is probably one we should also get @ChristianMcc 's thoughts on.

Reply:
This reply is just for @ChristianMcc to have some context.

Somewhat implicitly, the use of the fee record state machine here ensures that the fee record is in the appropriate state to add a payment. For instance, if the fee record is in `MATCH` state and you try to add another payment, that error would be caught here. The computation of which status to set is done at the utilisation report state machine level mostly for performance reasons, but also because there should really only be one fee record attached to one fee record state machine. The `'PAYMENT_ADDED'` event accepting the new status to set makes sense to me logically, as we are telling the fee record state machine that a new payment has been added and your payments now match or do not match. The state machine then says either "no, you can't add a new payment to this fee record" or "okay, I'll update the status"

Reply:
Ooh that is an interesting perspective, I had kind of forgotten about that state check on the state machine. 
As a side comment on that note, do we want to prevent them adding another payment if the fee record has match status? I can't see that AC on any ticket

Reply:
I was kind of in the mind that we should just treat fee records as groups but I have just realised that once we key they can be updated separately to done so I guess we shouldn't entertain that idea...
We could instead just tell the fee record state machine whether the payments match the reported payments via a boolean and then let the fee record state machine own the knowledge of what status to set? It's a shame it would be really inefficient to calculate it for each fee record inside the state machine...

Reply:
> As a side comment on that note, do we want to prevent them adding another payment if the fee record has match status? I can't see that AC on any ticket

This is less an "ought" comment and more an "is" from me - at present, the state machine does only permit adding a payment if you are in the `TO_DO` status. I am going to update it so that you can add a payment if the status is `TO_DO`, `DOES_NOT_MATCH` and `MATCH`